src/Umbraco.Web/Security/OwinDataProtectorTokenProvider.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Owin.Security.DataProtection;
+using Umbraco.Web.Models.Identity;
+
+namespace Umbraco.Web.Security
+{
+    /// <summary>
+    /// Adapted from Microsoft.AspNet.Identity.Owin.DataProtectorTokenProvider
+    /// </summary>
+    public class OwinDataProtectorTokenProvider<TUser> : IUserTwoFactorTokenProvider<TUser> where TUser : BackOfficeIdentityUser
+    {
+        public TimeSpan TokenLifespan { get; set; }
+        private static readonly Encoding _defaultEncoding = new UTF8Encoding(false, true);
+        private readonly IDataProtector _protector;
+
+        public OwinDataProtectorTokenProvider(IDataProtector protector)
+        {
+            _protector = protector ?? throw new ArgumentNullException(nameof(protector));
+            TokenLifespan = TimeSpan.FromDays(1);
+        }
+
+        public async Task<string> GenerateAsync(string purpose, UserManager<TUser> manager, TUser user)
+        {
+            if (manager == null) throw new ArgumentNullException(nameof(manager));
+            if (user == null) throw new ArgumentNullException(nameof(user));
+
+            var ms = new MemoryStream();
```

Comment:
Think guess we need to wrap this in a using statement to ensure is is disposed after the `ms.ToArray()`

Reply:
Agreed. I ported this directly from the MS implementation, so this is what you were using before. Let me see if there is a reason that they've left it open so that I don't somehow create a regression.

Reply:
I wouldn't worry too much about it, it will be deleted anyways :) If it's working in the back office now then I'd say it doesn't matter too much.

Reply:
https://stackoverflow.com/questions/234059/is-a-memory-leak-created-if-a-memorystream-in-net-is-not-closed/234071